pkg/devfile/adapters/kubernetes/component/podwatcher.go

```diff
@@ -0,0 +1,441 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog"
+)
+
+const (
+	// KubernetesResourceFailureInterval is the time between attempts to acquire needed k8s resources
+	KubernetesResourceFailureInterval = time.Duration(5) * time.Second
+)
+
+// podWatcher is responsible for watching for changes to odo-managed Pods, and reporting those changes to the console, as used by the status command
+type podWatcher struct {
+	adapter              *Adapter
+	statusReconcilerChan chan statusReconcilerChannelEntry
+}
+
+// StartContainerStatusWatch outputs Kubernetes pod/container status changes to the console, as used by the status command
+func (a Adapter) StartContainerStatusWatch() {
+
+	pw := newPodWatcher(&a)
+	pw.startPodWatcher()
+
+}
+
+func newPodWatcher(adapter *Adapter) *podWatcher {
+	return &podWatcher{
+		adapter:              adapter,
+		statusReconcilerChan: createStatusReconciler(adapter),
+	}
+}
+
+func (pw *podWatcher) startPodWatcher() {
+	pw.startWatchThread(pw.adapter)
+}
+
+// statusReconcilerChannelEntry is the gochannel message sent from the Watcher to the status reconciler
+type statusReconcilerChannelEntry struct {
+
+	// If isCompleteListOfPods is true: a list of all component pods in the workspace
+	// If isCompleteListOfPods is false: a single component pod in the workspace
+	pods []*corev1.Pod
+
+	err error
+
+	// isCompleteListOfPods is true if the pods came from getLatestContainerStatus(), false otherwise
+	isCompleteListOfPods bool
+
+	// isDeleteEventFromWatch is true if watch.Deleted event from the watch, false otherwise
+	isDeleteEventFromWatch bool
+
+	// watchThreadRestarted is true if the watch thread died (for example, due to losing network connection) and had to be reestablished
+	isWatchThreadRestarted bool
+}
+
+// getLatestContainerStatus returns a KubernetesDeploymentStatus for the given component; this function blocks until it is available
+func getLatestContainerStatus(adapter *Adapter) *KubernetesDeploymentStatus {
+
+	// Keep trying to acquire the ReplicaSet and DeploymentSet of the component, so that we can reliably find its pods
+	for {
+		containerStatus, err := adapter.getDeploymentStatus()
+		if err == nil {
+
+			if containerStatus.DeploymentUID == "" || containerStatus.ReplicaSetUID == "" {
+				adapter.Logger().ReportError(fmt.Errorf("unable to retrieve component deployment and replica set, trying again in a few moments"), machineoutput.TimestampNow())
+				time.Sleep(KubernetesResourceFailureInterval)
+				continue
+			}
+
+			return containerStatus
+		}
+
+		adapter.Logger().ReportError(errors.Wrapf(err, "unable to retrieve component deployment and replica set, trying again in a few moments"), machineoutput.TimestampNow())
+		time.Sleep(KubernetesResourceFailureInterval)
+	}
+
+}
+
+func (pw *podWatcher) startWatchThread(adapter *Adapter) {
+
+	// Kick off the goroutine then return execution
+	go func() {
+
+		watchAttempts := 0
```

Comment:
Nit: Why not initialize watchAttemps to 1? Then we don't need to do `watchAttempts+1` when we print it

Reply:
Done.